---
title: "05.Downstream_analysis"
author: "junyao"
date: "2024-05-25"
output: html_document
---
# Here is the example to run downstream analysis in DRCTdb.

## load packages

```r
library(Seurat)
library(igraph)
library(GenomicRanges)
source('E:\\DRCTdb\\5.Downstream_analysis/identify_region_motif.R')
source('E:\\DRCTdb\\5.Downstream_analysis/overlap_gwas.R')
source('E:\\DRCTdb\\5.Downstream_analysis/plot.R')
```

## set up data paths
```r
output_path = 'E:\\DRCTdb\\ignore\\downstream_result\\sample1\\'
rna_path = 'F:\\DRCTdb\\ignore\\scRNA-seq\\sample1\\sample1_heart_scRNA_35k_processed.Rds'
atac_path = 'E:\\DRCTdb\\ignore\\bed\\sample1/'
ldsc_path = "E:/DRCTdb/ignore/LDSC_results/sample1/pvalues.tsv"
snp_path = 'E:\\public\\all_snp_info_gr.Rds'
disease_path = 'E:\\DRCTdb\\ignore\\LDSC_hg38\\summary_statistics\\Josh'
```

## load data
```r
rna = readRDS(rna_path)
rna@active.ident=as.factor(rna$celltype)
atac = dir(atac_path)
atac_ct = unlist(strsplit(atac,'.bed.gz'))
ct_use = intersect(rna$celltype,atac_ct)
names(atac) = atac_ct
atac = atac[names(atac) %in% ct_use]
atac_list = list()
for (i in 1:length(atac)) {
  bed=read.delim(paste0(atac_path,atac[i]),header = F)
  peak = paste0(bed[,1],':',bed[,2],'-',bed[,3])
  atac_list[[names(atac)[i]]] = peak
}
ldsc_result <- read.delim(ldsc_path)
ct_use = intersect(colnames(ldsc_result),ct_use)
snp_all = readRDS(snp_path)
ldsc_result$X = gsub(' ','_',ldsc_result$X)
disease_all = dir(disease_path)
disease_all_name = as.data.frame(t(as.data.frame(strsplit(disease_all,'\\.'))))
names(disease_all) = disease_all_name[,4]
```


## Part1, disease related peaks, genes, and GRNs
we aim to determine the relationship between disease-associated single nucleotide polymorphisms (SNPs) and peaks identified from single-cell ATAC sequencing (scATAC), as well as genes identified from single-cell RNA sequencing (scRNA). Subsequently, we will construct a gene regulatory network (GRN) using the identified peaks and genes. To fastly demonstrate the codes, we just run this analysis in 2 cell types.
```r
grn_list02 = list()
snp_list = list()
for (i in 1:2) {
  ### define significant disease
  disease_use = ldsc_result[,1][ldsc_result[,ct_use[i]]<0.05]
  disease_use = unique(disease_use)
  disease_use = intersect(disease_use,names(disease_all))
  for (j in disease_use) {
      rna_use = subset(rna,celltype==ct_use[i])
      list1 = gwas_related_features(rna_use,atac_list[[i]],
                            disease_name=j,
                            snp_all = snp_all,zscore_thr = 1)

      grn04 = ct_grn_atac(list1[[2]][,1:3],
                        unique(list1[[1]]$symbol),
                        rna_use,cor_thr=0.4)
      grn02 = ct_grn_atac(list1[[2]][,1:3],
                        unique(list1[[1]]$symbol),
                        rna_use,cor_thr=0.2)
      grn_name = paste0(ct_use[i],'_',j)
      grn_list04[[grn_name]] = grn04
      grn_list02[[grn_name]] = grn02
      snp_list[[grn_name]] = list1
  }

}

```


## Part2, disease realted cell type summary
we summarise the ldsc analysis result, and identify disease related cell types by the LDSC pvalue.
```r
ldsc_result = ldsc_result[!duplicated(ldsc_result[,1]),]
rownames(ldsc_result) = ldsc_result[,1]
ldsc_result = ldsc_result[,ct_use]
sig_ct_list = list()
sig_ct_vector = c()
for (i in 1:nrow(ldsc_result)) {
  sig_ct = colnames(ldsc_result)[ldsc_result[i,]<0.05]
  if (length(sig_ct)>0) {
    sig_ct_list[[rownames(ldsc_result)[i]]] = sig_ct
    sig_ct_merge = paste0(sig_ct,collapse = ';')
    sig_ct_vector = c(sig_ct_vector,sig_ct_merge)
  }
}
sig_ct_df = data.frame(names(sig_ct_list),sig_ct_vector)
colnames(sig_ct_df) = c('disease','related_cell_type')
print(sig_ct_df)
```

## Part3, cell-cell communications among disease related cell types
```r
### disease related ccc
ccc_plot_list = list()
ccc_list = list()
source('F:\\general_code\\run_cellchat.R')
for (i in 1:length(sig_ct_list)) {
  disease_name_use = names(sig_ct_list)[i]
  related_ct = sig_ct_list[[i]]
  if (length(related_ct)>1) {
    rna_use = subset(rna,celltype %in% related_ct)
    ccc = run_cellchat(rna_use,rna_use@meta.data,group = 'celltype',species = 'hs')
    groupSize <- as.numeric(table(ccc@idents))
    par(mfrow=c(1,1))
    p1 = netVisual_circle(ccc@net$weight, vertex.weight = groupSize,
                          weight.scale = T, label.edge= F,
                          title.name = "",vertex.label.cex = 1)
    print(disease_name_use)
    ### disease related ccc
    ccc_plot_list[[i]] = p1
    ccc_list[[i]] = ccc
  }
}
```


